### ISSUE TITLE
[Bug]: [laion1b-test] GC has not been triggered for a long time

### ISSUE THREAD

**Issue by ThreadDao** [milvus-io/milvus#29655]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version: 2.3-20231229-7a192da8-amd64
- Deployment mode(standalone or cluster): cluster
- MQ type(rocksmq, pulsar or kafka):   pulsar 
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

1.  Concurrent insert, delete, flush, query, search of a collection with 50 million data. The number of segments in the collection increases sharply and then stops increasing new data. [laion1b-test-read-write-new-8a](https://argo-workflows.zilliz.cc/archived-workflows/qa/7843d2e1-6113-4ff1-8e65-d47e6edaf5ba?nodeId=laion1b-test-read-write-new-8a)

2. After a period of index building and compaction, the number of segments needed is reduced. But the number of dropped segments has not decreased. I guess GC is blocked by something
![image](https://github.com/milvus-io/milvus/assets/27288593/c25a5f3a-f03e-4482-8387-ded57794cf22)

[grafana metrics](https://grafana-4am.zilliz.cc/d/uLf5cJ3Ga/milvus2-0?orgId=1&var-datasource=prometheus&var-cluster=&var-namespace=qa-milvus&var-instance=laion1b-test-2&var-collection=All&var-app_name=milvus&from=1704180240000&to=1704254400000)

**Comment by SimFG** [comment]
@ThreadDao 
you can try to use the latest 2.3 branch to build a image, it should have been fixed.
link pr: https://github.com/milvus-io/milvus/pull/29557

**Comment by yanliang567** [comment]
/assign @ThreadDao 
please help to verify
/unassign

**Comment by ThreadDao** [comment]
@SimFG 
I thought the problem with this issue was that something might be blocking the GC rather than the lack of concurrency in the GC？

**Comment by SimFG** [comment]
@ThreadDao This is caused by too many dropped segments, which means the cleaning speed is relatively slow and it blocks the GC

**Comment by SimFG** [comment]
@ThreadDao @chyezh 
From the server log, i found it maybe block the recycle unused indexes step.
[loglink](https://grafana-4am.zilliz.cc/explore?orgId=1&left=%7B%22datasource%22:%22Loki%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bnamespace%3D%5C%22qa-milvus%5C%22,pod%3D~%5C%22laion1b-test-2-milvus-mixcoord-.*%5C%22%7D%20%7C~%20%5C%22start%20recycleUnusedIndexes%7Cstart%20recycleUnusedIndexFiles%7Cgarbage%20collector%20paused%7CsegmentID%20is%20not%20in%20drops%7Cskipping%20GC%7Cdropped%20segment%20dml%20position%20after%20channel%20cp%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%221704151440000%22,%22to%22:%221704225600000%22%7D%7D)

**Comment by chyezh** [comment]
GC Task blocks at scan (list) minio object.

```
goroutine 7744 [chan receive]:
github.com/milvus-io/milvus/internal/storage.(*MinioObjectStorage).ListObjects(0xc0008dc280, {0x555e678, 0xc02db23720}, {0xc000418770, 0xd}, {0xc02d1a6be0, 0x10}, 0x1)
	/go/src/github.com/milvus-io/milvus/internal/storage/minio_object_storage.go:190 +0x525
github.com/milvus-io/milvus/internal/storage.(*RemoteChunkManager).listObjects(0xc0008eb530, {0x555e678, 0xc02db23720}, {0xc000418770, 0xd}, {0xc02d1a6be0, 0x10}, 0x37?)
	/go/src/github.com/milvus-io/milvus/internal/storage/remote_chunk_manager.go:393 +0xca
github.com/milvus-io/milvus/internal/storage.(*RemoteChunkManager).ListWithPrefix(0xc079463768?, {0x555e678?, 0xc02db23720?}, {0xc02d1a6be0?, 0xc02be38f30?}, 0x2c?)
	/go/src/github.com/milvus-io/milvus/internal/storage/remote_chunk_manager.go:321 +0x3f
github.com/milvus-io/milvus/internal/datacoord.(*garbageCollector).scan(0xc059518fc0)
	/go/src/github.com/milvus-io/milvus/internal/datacoord/garbage_collector.go:218 +0x39c
github.com/milvus-io/milvus/internal/datacoord.(*garbageCollector).work(0xc059518fc0)
	/go/src/github.com/milvus-io/milvus/internal/datacoord/garbage_collector.go:153 +0xad9
created by github.com/milvus-io/milvus/internal/datacoord.(*garbageCollector).start.func1
	/go/src/github.com/milvus-io/milvus/internal/datacoord/garbage_collector.go:96 +0x6d
```

**Comment by chyezh** [comment]
> GC Task blocks at scan (list) minio object.
> 
> ```
> goroutine 7744 [chan receive]:
> github.com/milvus-io/milvus/internal/storage.(*MinioObjectStorage).ListObjects(0xc0008dc280, {0x555e678, 0xc02db23720}, {0xc000418770, 0xd}, {0xc02d1a6be0, 0x10}, 0x1)
> 	/go/src/github.com/milvus-io/milvus/internal/storage/minio_object_storage.go:190 +0x525
> github.com/milvus-io/milvus/internal/storage.(*RemoteChunkManager).listObjects(0xc0008eb530, {0x555e678, 0xc02db23720}, {0xc000418770, 0xd}, {0xc02d1a6be0, 0x10}, 0x37?)
> 	/go/src/github.com/milvus-io/milvus/internal/storage/remote_chunk_manager.go:393 +0xca
> github.com/milvus-io/milvus/internal/storage.(*RemoteChunkManager).ListWithPrefix(0xc079463768?, {0x555e678?, 0xc02db23720?}, {0xc02d1a6be0?, 0xc02be38f30?}, 0x2c?)
> 	/go/src/github.com/milvus-io/milvus/internal/storage/remote_chunk_manager.go:321 +0x3f
> github.com/milvus-io/milvus/internal/datacoord.(*garbageCollector).scan(0xc059518fc0)
> 	/go/src/github.com/milvus-io/milvus/internal/datacoord/garbage_collector.go:218 +0x39c
> github.com/milvus-io/milvus/internal/datacoord.(*garbageCollector).work(0xc059518fc0)
> 	/go/src/github.com/milvus-io/milvus/internal/datacoord/garbage_collector.go:153 +0xad9
> created by github.com/milvus-io/milvus/internal/datacoord.(*garbageCollector).start.func1
> 	/go/src/github.com/milvus-io/milvus/internal/datacoord/garbage_collector.go:96 +0x6d
> ```

ListObject of huge bucket in minio cost too much time.

**Comment by xiaofan-luan** [comment]
Let's take this in the newer fix

**Comment by chyezh** [comment]
New fix has been merged into master.
@ThreadDao Please verify it.
